rflx/pyrflx/typevalue.py

```diff
@@ -727,47 +767,139 @@ def set_refinement(fld: MessageValue.Field, fld_name: str) -> None:
+        if checksum_field_name not in self.fields:
+            raise KeyError(
+                f"cannot set checksum function: field {checksum_field_name} is not defined"
+            )
+        for field_name, checksum in self._checksums.items():
+            if field_name == checksum_field_name:
+                checksum.set_checksum_function(checksum_method)
+                return
+        raise KeyError(
+            f"cannot set checksum function: field {checksum_field_name} "
+            f"has not been defined as a checksum field"
+        )
+
+    def _is_checksum_settable(self, checksum: "MessageValue.Checksum") -> bool:
+        def valid_path(value_range: ValueRange) -> bool:
+            expr: Dict[Expr, str] = dict.fromkeys([value_range.lower, value_range.upper])
+
+            for e in expr:
+                if isinstance(e, Sub):
+                    assert isinstance(e.left, (First, Last))
+                    expr[e] = str(e.left.prefix)
+                elif isinstance(e, Add):
+                    for t in e.terms:
+                        if isinstance(t, (First, Last)):
+                            expr[e] = str(t.prefix)
+                else:
+                    assert isinstance(e, (First, Last))
+                    expr[e] = str(e.prefix)
+
+            field = expr.get(value_range.lower)
+            assert isinstance(field, str)
+            upper_field_name = expr[value_range.upper]
+            if upper_field_name == "Message":
+                upper_field_name = "Final"
+            while field != upper_field_name:
+                field = self._next_field(field)
+                if field == "Final":
+                    continue
+                if field == "" or not self._fields[field].set:
+                    break
+            else:
+                return True
+
+            return False
+
+        for expr_tuple in checksum.parameters:
+            expr_tuple.evaluated_expression = self.__simplified(expr_tuple.expression)
+            if (
+                isinstance(expr_tuple.evaluated_expression, ValueRange)
+                and isinstance(expr_tuple.expression, ValueRange)
+                and (
+                    not isinstance(expr_tuple.evaluated_expression.lower, Number)
+                    or not isinstance(expr_tuple.evaluated_expression.upper, Number)
+                    or not valid_path(expr_tuple.expression)
+                )
+            ):
+                break
+            if (
+                isinstance(expr_tuple.evaluated_expression, Variable)
+                and not self._fields[expr_tuple.evaluated_expression.name].set
+            ):
+                break
+            if (
+                isinstance(expr_tuple.evaluated_expression, Attribute)
+                and not self._fields[str(expr_tuple.evaluated_expression.prefix)].set
+            ):
+                break
+        else:
+            return True
+
+        return False
```

Comment:
I think the use of for else can be misleading here. It would be clearer if we would return False instead of break and return True at the end of the function. Also is there any reason to split the if conditions?

Reply:
Changed in 68261a1